internals/controllers/migrationhierarchy/migrationhierarchy_controller.go

```diff
@@ -0,0 +1,440 @@
+// +kubebuilder:rbac:groups=dana.hns.io,resources=migrationhierarchies/status,verbs=get;update;patch
+
+const maxRetries = 600
+
+func (r *MigrationHierarchyReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&danav1.MigrationHierarchy{}).
+		Complete(r)
+}
+
+func (r *MigrationHierarchyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithName("controllers").WithName("MigrationHierarchy").WithValues("mh", req.NamespacedName)
+	logger.Info("starting to reconcile")
+
+	mhObject, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Name: req.NamespacedName.Name}, &danav1.MigrationHierarchy{})
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get object '%s': "+err.Error(), mhObject.Object.GetName())
+	}
+
+	if !mhObject.IsPresent() {
+		logger.Info("resource not found. Ignoring since object must be deleted")
+		return ctrl.Result{}, nil
+	}
+
+	phase := mhObject.Object.(*danav1.MigrationHierarchy).Status.Phase
+	if utils.ShouldReconcile(phase) {
+		return ctrl.Result{}, r.reconcile(mhObject)
+	} else {
+		logger.Info("no need to reconcile, object phase is: ", "phase", phase)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *MigrationHierarchyReconciler) reconcile(mhObject *utils.ObjectContext) error {
+	ctx := mhObject.Ctx
+	logger := log.FromContext(ctx)
+
+	currentNamespace := mhObject.Object.(*danav1.MigrationHierarchy).Spec.CurrentNamespace
+	toNamespace := mhObject.Object.(*danav1.MigrationHierarchy).Spec.ToNamespace
+
+	// get the namespace of the subnamespace that should be migrated
+	ns, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Namespace: "", Name: currentNamespace}, &corev1.Namespace{})
+	if err != nil {
+		return fmt.Errorf("failed getting namespace object '%s': "+err.Error(), currentNamespace)
+	}
+
+	sourceSNSParentName := utils.GetNamespaceParent(ns.Object)
+	oldSNS, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Name: currentNamespace, Namespace: sourceSNSParentName}, &danav1.Subnamespace{})
+	if err != nil {
+		return fmt.Errorf("failed getting subnamespace object '%s': "+err.Error(), sourceSNSParentName)
+	}
+
+	// create a new subnamespace under the new parent
+	newSNS, reason, er := r.createSNS(oldSNS, currentNamespace, toNamespace)
+	if er != nil {
+		err := r.updateMHErrorStatus(mhObject, er, reason)
+		if err != nil {
+			return fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+		}
+		return fmt.Errorf("failed creating subnamespace '%s' under namespace '%s': "+er.Error(), currentNamespace, toNamespace)
+	}
+	logger.Info("successfully created new subnamespace under new parent", "subnamespace", currentNamespace, "new parent", toNamespace)
+
+	// delete the old subnamespace under the original parent
```

Comment:
wrap in function delete old namespace and i would create inside another function roll back function

Reply:
not sure another rollback function is needed here